example/stochastic_subgrid/stochastic_subgrid_package.cpp

```diff
@@ -0,0 +1,462 @@
+  if ((pin->GetOrAddInteger("parthenon/mesh", "nx3", 1) > 1) && ang_2 != 0.0)
+    lambda = std::min(lambda, x3);
+
+  // If cos_a2 or cos_a3 = 0, need to override lambda
+  if (ang_3_vert) lambda = x2;
+  if (ang_2_vert) lambda = x3;
+
+  // Initialize k_parallel
+  Real k_par = 2.0 * (M_PI) / lambda;
+
+  pkg->AddParam<>("amp", amp);
+  pkg->AddParam<>("vel", vel);
+  pkg->AddParam<>("vx", vx);
+  pkg->AddParam<>("vy", vy);
+  pkg->AddParam<>("vz", vz);
+  pkg->AddParam<>("k_par", k_par);
+  pkg->AddParam<>("cos_a2", cos_a2);
+  pkg->AddParam<>("cos_a3", cos_a3);
+  pkg->AddParam<>("sin_a2", sin_a2);
+  pkg->AddParam<>("sin_a3", sin_a3);
+
+  // get parameters for powerlaw distribution to sample number of iterations from
+  int N_min = pin->GetOrAddInteger("Random", "num_iter_min", 1);
+  int N_max = pin->GetOrAddInteger("Random", "num_iter_max", 100);
+  Real alpha = pin->GetOrAddReal("Random", "power_law_coeff", -3.0);
+
+  if (N_min <= 0) PARTHENON_FAIL("Random/num_iter_min must be > 0");
+  if (N_max < N_min) PARTHENON_FAIL("Random/num_iter_max must be >= Random/num_iter_min");
+
+  int N = N_max - N_min + 1;
+
+  Kokkos::View<int *> num_iter_hist("num_iter_histogram", N);
+  auto num_iter_hist_host = Kokkos::create_mirror_view(num_iter_hist);
+
+  pkg->AddParam("num_iter_histogram", num_iter_hist);
+  pkg->AddParam("N_min", N_min);
+
+  // compute non-normalized probabilities
+  std::vector<Real> prob(N);
+  for (int i = 0; i < N; ++i) {
+    prob[i] = pow(i + N_min, alpha);
+    num_iter_hist_host(i) = 0;
+  }
+
+  AliasMethod alias(prob);
+  pkg->AddParam("alias_method", alias);
+
+  Kokkos::deep_copy(num_iter_hist, num_iter_hist_host);
+
+  // create random pool
+  uint64_t seed = pin->GetOrAddInteger("Random", "seed", 0);
+  // if we don't have a seed, use the time
+  if (seed == 0)
+    seed = std::chrono::high_resolution_clock::now().time_since_epoch().count();
+
+  Kokkos::Random_XorShift64_Pool<parthenon::DevExecSpace> rand_pool(seed);
+  pkg->AddParam("random_pool", rand_pool);
+
+  // number of variable in variable vector
+  const auto num_vars = pin->GetOrAddInteger("Advection", "num_vars", 1);
+
+  std::string field_name = "advected";
+  Metadata m({Metadata::Cell, Metadata::Independent, Metadata::FillGhost},
+             std::vector<int>({num_vars}));
+  pkg->AddField(field_name, m);
+
+  field_name = "dummy_result";
+  m = Metadata({Metadata::Cell, Metadata::Derived, Metadata::OneCopy});
+  pkg->AddField(field_name, m);
+
+  field_name = "num_iter";
+  m = Metadata({Metadata::Cell, Metadata::Derived, Metadata::OneCopy});
+  pkg->AddField(field_name, m);
+
+  pkg->FillDerivedBlock = DoLotsOfWork;
+  pkg->CheckRefinementBlock = CheckRefinement;
+  pkg->EstimateTimestepBlock = EstimateTimestepBlock;
+
+  return pkg;
+}
+
+AmrTag CheckRefinement(MeshBlockData<Real> *rc) {
+  auto pmb = rc->GetBlockPointer();
+  // refine on advected, for example.  could also be a derived quantity
+  auto v = rc->Get("advected").data;
```

Comment:
Can `v` be made const reference?

Reply:
Hmm, good question. I think it could, but I don't think we gain anything. It goes into a `KOKKOS_LAMBDA` that captures it by value and thus makes a copy and I think the Kokkos documentation encourages to use `view` by value since it's very lightweight and can do reference counting properly that way.